New translation: creating-and-modifying-pages.md (#174)
gatsbyjs/gatsby-pt-BR: docs/docs/creating-and-modifying-pages.md

---
title: Criando e modificando páginas
---

O Gatsby facilita a manipulação pragramática das suas páginas.

As páginas podem ser criadas de três maneiras:

- Através da implementação da API [`createPages`](/docs/node-apis/#createPages) no arquivo gatsby-node.js do seu site.
- O Gatsby core automaticamente transforma os componentes React nas `src/pages` em páginas.

> - Plugins também pode implementar o `createPages` e crias as páginas para você.

Reviewer: ```suggestion
- Plugins também pode implementar o `createPages` e criar páginas para você.
```

> Você também pode implementar a API [`onCreatePage`](/docs/node-apis/#onCreatePage) caso queira modificar páginas criadas no Gatsby core, por plugins ou também criar [rotas-apenas-para-clientes](/docs/building-apps-with-gatsby/).

Reviewer: ```suggestion
Você também pode implementar a API [`onCreatePage`](/docs/node-apis/#onCreatePage) caso queira modificar páginas criadas no Gatsby core, por plugins ou também criar [rotas apenas-para-clientes](/docs/building-apps-with-gatsby/).
```

## Ajuda na depuração

> Para ver quais páginas estão sendo criadas pelo seu código ou plugins, você pode consultar as  informações da página enquanto desenvolve no Graph*i*QL. Cole a seguinte consulta na IDE do GraphiQL para o seu site. A IDE do Graph*i*QL fica disponível ao executar o server de desenvolvimento do seu site na `HOST:PORT/___graphql` e.g.

Reviewer: ```suggestion
Para ver quais páginas estão sendo criadas pelo seu código ou plugins, você pode consultar as informações da página enquanto desenvolve no Graph*i*QL. Cole a seguinte consulta na IDE do Graph*i*QL para o seu site. A IDE do Graph*i*QL fica disponível ao executar o servidor de desenvolvimento do seu site na `HOST:PORT/___graphql` e.g.
```

}
```

> A propriedade `context` aceita objetos, então, podemos passar quaisquer dados que nós queiramos que a página possa acessar. 

Reviewer: ```suggestion
A propriedade `context` aceita objetos, então podemos passar quaisquer dados que nós queiramos que a página possa acessar. 
```

> Você também pode consultar quaisquer dados `context` adicionadas por você, ou por plugins utilizados.

Reviewer: ```suggestion
Você também pode consultar quaisquer dados `context` adicionadas por você ou pelos plugins utilizados.
```

> > **NOTE:** Existem algumas palavras reservadas que _não_ podem ser utilizadas no `context`. As palavras são: `path`, `matchPath`, `component`, `componentChunkName`, `pluginCreator___NODE`, e `pluginCreatorId`.

Reviewer 2: ```suggestion
> **NOTA:** Existem algumas palavras reservadas que _não_ podem ser utilizadas no `context`. As palavras são: `path`, `matchPath`, `component`, `componentChunkName`, `pluginCreator___NODE`, e `pluginCreatorId`.
```

## Criando páginas no gatsby-node.js

Frequentemente, você precisará criar páginas através de código. Por exemplo, você terá arquivos markdown, onde cada um deve ser uma página.

> Esse exemplo assume que cada página markdown tem um `path` definido no frontmatter do arquivo markdown.

Reviewer: ```suggestion
Esse exemplo assume que cada página markdown tem um `path` definido no _frontmatter_ do arquivo markdown.
```

}
```

## Modificando páginas criadas pelo Gatsby core ou por plugins

Tanto o Gatsby core quanto plugins podem criar páginas para você de forma automática. Algumas vezes o padrão criado não é exatamente o que você quer e, por isso, você precisará modificar essas páginas criadas.

### Removendo barras à direita

Uma razão comum para modificar automaticamente páginas já criadas é a remoção de barras à direita

Para isso, no arquivo `gatsby-node.js` do seu site adicione um código similar ao seguinte:

> _Nota: Existe um plugin que irá remover todos as barras à direita da página automaticamente:

Reviewer: ```suggestion
_Nota: Existe um plugin que irá remover todos as barras à direita das páginas automaticamente:
```

[plugin-gatsby-que-remove-as-barras-à-direita](/packages/gatsby-plugin-remove-trailing-slashes/)_.

> _Nota: Caso você precise fazer uma ação assíncrona no `onCreatePage` você pode retornar uma promessa ou usar uma função `async`._

Reviewer: ```suggestion
_Nota: Caso você precise fazer uma ação assíncrona no `onCreatePage` você pode retornar uma _promise_ ou usar uma função `async`._
```

```javascript:title=gatsby-node.js
// substituir o '/' resultaria em uma string vazia, o que é inválido.
const replacePath = path => (path === `/` ? path : path.replace(/\/$/, ``))
// Implemente a API do Gatsby “onCreatePage”. Isso é
// executado logo que uma página é criada
exports.onCreatePage = ({ page, actions }) => {
  const { createPage, deletePage } = actions

  const oldPage = Object.assign({}, page)
  // Remove a barra à direita a não ser que a página seja /
  page.path = replacePath(page.path)
  if (page.path !== oldPage.path) {
    // substitui a página antiga pela nova
    deletePage(oldPage)
    createPage(page)
  }
}
```

### Adicionar contexto à páginas

> As páginas criadas automaticamente podem receber contextos e usa-lo como variáveis nas consultas da GraphQL. Para substituir o padrã e usar o seu próprio contexto, abra o arquivo `gatsby-node.js` do seu site e adicione um código similar ao seguinte:

Reviewer: ```suggestion
As páginas criadas automaticamente podem receber contextos e usa-lo como variáveis nas consultas da GraphQL. Para substituir o padrão e usar o seu próprio contexto, abra o arquivo `gatsby-node.js` do seu site e adicione um código similar ao seguinte:
```

> As páginas criadas automaticamente podem receber contextos e usa-lo como variáveis nas consultas da GraphQL. Para substituir o padrão e usar o seu próprio contexto, abra o arquivo `gatsby-node.js` do seu site e adicione um código similar ao seguinte:

Reviewer 2: ```suggestion
As páginas criadas automaticamente podem receber contextos e usá-lo como variáveis nas consultas da GraphQL. Para substituir o padrão e usar o seu próprio contexto, abra o arquivo `gatsby-node.js` do seu site e adicione um código similar ao seguinte:
```

export default Page
```

> O contexto da página é serializado antes de ser passada para as páginas: Isso significa que ela não pode ser usado para transformar funções em componentes

Reviewer: ```suggestion
O contexto da página é serializado antes de ser passada para as páginas: Isso significa que ela não pode ser usada para transformar funções em componentes
```

> O contexto da página é serializado antes de ser passada para as páginas: Isso significa que ela não pode ser usada para transformar funções em componentes

Reviewer 2: ```suggestion
O contexto da página é serializado antes de ser passado para as páginas: Isso significa que ele não pode ser usado para transformar funções em componentes
```

## Criando rotas somente para clientes

> Em casos específicos, talvez você queira criar sites com porções apenas para clientes que são bloqueadas por autenticação. Para saber como faze-lo, recorra ao [Rotas-apenas-para-clientes & autenticação-de-usuário](https://www.gatsbyjs.org/docs/client-only-routes-and-user-authentication/).

Reviewer: ```suggestion
Em casos específicos, talvez você queira criar sites com porções apenas para clientes que são bloqueadas por autenticação. Para saber como faze-lo, recorra ao [Rotas apenas-para-clientes & autenticação de usuário](https://www.gatsbyjs.org/docs/client-only-routes-and-user-authentication/).
```

> Em casos específicos, talvez você queira criar sites com porções apenas para clientes que são bloqueadas por autenticação. Para saber como faze-lo, recorra ao [Rotas apenas-para-clientes & autenticação de usuário](https://www.gatsbyjs.org/docs/client-only-routes-and-user-authentication/).

Reviewer 2: ```suggestion
Em casos específicos, talvez você queira criar sites com porções apenas para clientes que são bloqueadas por autenticação. Para saber como fazê-lo, recorra ao [Rotas apenas-para-clientes & autenticação de usuário](https://www.gatsbyjs.org/docs/client-only-routes-and-user-authentication/).
```
